pennylane/devices/qubit/preprocess.py

```diff
@@ -77,6 +78,76 @@ def _operator_decomposition_gen(
 #######################
 
 
+def validate_and_expand_adjoint(
+    circuit: qml.tape.QuantumScript,
+) -> qml.tape.QuantumScript:  # pylint: disable=protected-access
+    """Function for validating that the operations and observables present in the input circuit
+    are valid for adjoint differentiation.
+
+    Args:
+        circuit(.QuantumTape): the tape to validate
+
+    Returns:
+        .QuantumTape: The expanded tape, such that it is supported by adjoint differentiation
+
+    Raises:
+        QuantumFunctionError: if the circuit is invalid for adjoint differentiation
+    """
+    # Check validity of measurements
+    measurements = []
+    for m in circuit.measurements:
+        if not isinstance(m, ExpectationMP):
+            raise qml.QuantumFunctionError(
+                "Adjoint differentiation method does not support"
+                f" measurement {m.__class__.__name__}"
+            )
+
+        if m.obs.name in {"Hamiltonian"}:
+            raise qml.QuantumFunctionError(
+                f"Adjoint differentiation method does not support observable {m.obs.name}."
+            )
+
+        if not hasattr(m.obs, "base_name"):
+            m.obs.base_name = None  # This is needed for when the observable is a tensor product
```

Comment:
Still not sure where `base_name` is getting used.

Reply:
Matt's suggestion for checking if the observable has a matrix fixed the issue, so we don't need `base_name` anymore.